---
layout: research2
title: Data analysis and integration across cytometry and single-cell technologies
permalink: /research/analysis/
---

### High-dimensional cytometry and single-cell technologies

<div class='row'>
    <div class="image">
        <a href="#">
            <img src="https://images.squarespace-cdn.com/content/v1/4f5013cf24ac3afbd1aa0384/1455230070354-2VUW5KGW5XS42AGTAKEU/ke17ZwdGBToddI8pDm48kKV0nEQLrZCZyN20-sdpYlF7gQa3H78H3Y0txjaiv_0fDoOvxcdMmMKkDsyUqMSsMWxHk725yiiHCCLfrh8O1z4YTzHvnKhyp6Da-NYroOW3ZGjoBKy3azqku80C789l0ivq7Q1ckvJa8MA8qNUlEOZ5IGU-1O_EUPktdRWJqBeswtSHa-_ZlYvzXGRIA25Fyg/_YD_5607.jpg" alt="icon" width="300" align="left" style="padding-left: 0px; padding-right: 10px; padding-top: 5px; padding-bottom: 10px">
        </a>
    </div>
</div>

Critical to investigating complex immune responses to disease is the capacity to measure cells at the single-cell level, examining multiple features of the cell simultaneously. To address this, we develop and apply a variety of high-dimensional cytometry, single-cell, and spatial methods, through the [Sydney Cytometry Core Research Facility](https://sydneycytometry.org.au/). These technologies include high-dimensional [flow cytometry](https://immunedynamics.io/resources/flow) (including the world's first 10-laser 30-parameter ‘LSR-X’ platform developed with BD ([Ashhurst et al. 2017](https://currentprotocols.onlinelibrary.wiley.com/doi/abs/10.1002/cpim.37)), [spectral cytometry](https://immunedynamics.io/resources/flow), as well as Australia's first [mass cytometer](https://tomashhurst.github.io/resources/mass) (CyTOF) and [Imaging Mass Cytometer](https://tomashhurst.github.io/resources/imc) (IMC). Most recently this has expanded to including [single-cell transcriptomics and multi-omics](https://immunedynamics.io/resources/).

<br />
<br />

### Challenges in analysing high-dimensional data

<div class='row'>
    <div class="image">
        <a href="#">
            <img src="https://wiki.centenary.org.au/download/attachments/186841491/image2020-8-20_15-17-53.png?version=1&modificationDate=1613891307741&api=v2" alt="icon" width="300" align="left" style="padding-left: 0px; padding-right: 10px; padding-top: 5px; padding-bottom: 10px">
        </a>
    </div>
</div>

Key to a 'systems immunology' approach to studing inflammation and infectious disease is the capacity to  analyse and integrate datasets generated by different single-cell and cytometry technologies. In particular, the challenges in analysing high-dimensional flow, spectral, or mass cytometry (CyTOF) datasets are different to those typically faced in analysing single-cell sequencing data. This is largely due to the large number of cells in cytometry datasets, consisting of tens to hundreds of millions of cells. However, fewer features can be measured on individual cells through these cytometry technologies (typically <50 proteins), when compared to single-cell multiomic approaches (such as CITE-seq: thousands of transcripts and hundreds of proteins), though fewer cells are measured. Ideally, we could utilise a system that is versatile enough to manage data generated through different cytometry or single-cell technologies, with a built in capacity to integrate datasets across batches, experiments, or technologies.

<br />

### Analysis of large cytometry and single-cell datasets with Spectre

<div class='row'>
    <div class="image">
        <a href="#">
            <img src="https://wiki.centenary.org.au/download/attachments/146080604/SPECTRE?version=6&modificationDate=1588780080372&api=v2" alt="icon" width="300" align="left" style="padding-left: 0px; padding-right: 10px; padding-top: 5px; padding-bottom: 10px">
        </a>
    </div>
</div>

- A major focus of our collaborative group is the development of computational analysis solutions for high-dimensional cytometry data.

To address these challenges, we developed '[Spectre](https://immunedynamics.io/spectre/)', a computational toolkit in R that enables comprehensive end-to-end integration, exploration, and analysis of high-dimensional cytometry data from different batches or experiments ([Ashhurst et al, 2021](https://www.biorxiv.org/content/10.1101/2020.10.22.349563v1.abstract)). Spectre streamlines the analytical stages of raw data pre-processing, batch alignment, data integration, clustering, dimensionality reduction, visualisation and population labelling, as well as quantitative and statistical analysis; with a simple, clear, and modular design of analysis workflows, that can be utilised by data and laboratory scientists. 

<br />

### Data integration and classification

<div class='row'>
    <div class="image">
        <a href="#">
            <img src="https://wiki.centenary.org.au/download/attachments/186841491/image2021-2-28_12-3-31.png?version=1&modificationDate=1614474211754&api=v2" alt="icon" width="300" align="left" style="padding-left: 0px; padding-right: 10px; padding-top: 5px; padding-bottom: 10px">
        </a>
    </div>
</div>

Inspired by the efforts of the [Human Cell Atlas (HCA)](http://humancellatlas.org/), we sought to create an interface that would allow a seamless transition across a wide variety of single-cell data types - including flow, spectral, CyTOF, and scRNAseq, including multiomic assays such as CITE-seq and Abseq. In particular, our goal was to faciliate integration and label transfer between datasets, so that reference datasets generated by the HCA and other groups could be leveraged in a wide variety of contexts, in both HD cytometry and single-cell sequencing. By creating a technology and data-agnostic platform, we can efficiently incorporate various analysis tools, and particular integration tools, developed for cytometry (e.g. FlowSOM, CytoNorm) and transcriptomics (e.g. Harmony, Liger, Seurat). By doing this, multiple approaches can be efficiently run and compared within a single environment. Critically, as many existing scRNAseq tools [do not scale well to large datasets](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9#Abs1), we have focused our analysis and integration approaches around methods that can scale to large datasets consisting of millions of cells.

Strategic implementation of batch-alignment, data-integration, and cell-type classification tools allow for the integrated analysis of multiple experiments, as well as a reproducible system for rapid and repeated cell type identification in large datasets. Critically, the fundamental data structures used within Spectre, along with the implementation of classifiers allow for the scalable analysis of very large high-dimensional datasets. In addition to high-dimensional cytometry datasets, we’ve also developed functions to allow for spatial analysis of high-dimensional imaging datasets, such as those generated by Imaging Mass Cytometry. 

Development
- Spectre for computational analysis
- Batch alignment/integration
- Paretobench and DR dev
- ICB paper example

Application
— examples in WNV/ZIKV/COVID etc

<br />
