---
layout: research2
title: Research Overview - Single-Cell Systems Immunology of Inflammation and Infectious Disease
permalink: /research/
---

---

**Background**: 

_SCSI_

- inflammation and infectious disease
- immunopathology in particular
- investigation of the immune response at the single cell level is critical to elucidating the mechanistic drivers of health and disease 

_TECH_

- the expansion of high-dimensional cytometry, single-cell, and imaging technologies has lead to a significant shift in our approach to analysing immunological data. 

_DATA_

- As the size and complexity of high-dimensional cytometry data continue to expand, comprehensive, scalable, and methodical computational analysis approaches are essential. 
- Yet, contemporary clustering and dimensionality reduction tools alone are insufficient to analyze or reproduce analyses across large numbers of samples, batches, or experiments. 
- Moreover, approaches that allow for the integration of data across batches or experiments are not well incorporated into computational toolkits to allow for streamlined workflows, nor are approaches to analyse time-series data. 

--> something about capacity to contribute to, or utilise, single-cell atlas datasets, such as those developed by the HCA, HubMap, and LifeTime.

<br />

**Our work**: 

Initiated through a seed grant from the [Marie Bashir Institute for Infectious Diseases and Biosecurity](https://www.sydney.edu.au/marie-bashir-institute/), we established the the [Immune Dynamics team](https://immunedynamics.io/team), a multi-disciplinary collaboration between multiple [research and technology labs](https://immunedynamics.io/team) at the University of Sydney; integrating expertise in immunology, infectious disease, single-cell & imaging technologies, computational biology, and data science. Our team seeks to comprehensively map dynamic immune repsonses at the single cell level, over time, in various inflammatory and infectious diseases. Through the development of novel [single-cell and imaging methods](), as well as [advanced computational analysis approaches](), we seek to address key challenges in the single-cell study of inflammation, and develop single-cell 'atlases' that may be leveraged to reveal the key mediators of health and disease, as part of larger efforts such as the [Human Cell Atlas]() project. Our team has a particular interest in the study of emerging pathogens, such as [SARS-CoV-2 and COVID-19](https://tomashhurst.github.io/research/#application-to-disease), [Zika virus](https://tomashhurst.github.io/research/#application-to-disease) encephalitis, and [West Nile virus](https://tomashhurst.github.io/research/#application-to-disease) encephalitis. 

Below we detail our specific areas of research and development. More information on the research interests of our individual members can be found on the [team](https://immunedynamics.io/team/) page.

<br />

## Single-cell systems immunology of inflammation and infectious disease

---

- Inflammatory response maps and atlases across diseases
- Integration approaches to leverage multiple single-cell and cytometry technologies
- COVID-19
- Viral encephalitis

<br />

## Data analysis and integration across cytometry and single-cell technologies

---

A major focus of our collaborative group is the development of computational analysis solutions for high-dimensional cytometry data.

- Important -- ability to test and compare different clustering, DR, and alignment/integration tools -- in Spectre, make it easy to run multiple tools to optimiset he approach 
- Inspire by efforts of HCA
- Sought to create an interface that would allow a seamless transition across a wide variety of single-cell data types -- including flow, spectral, CyTOF, and scRNAseq, including multiomic assays such as CITE-seq and Abseq
- In particular faciliate integration and label transfer between datasets, so that reference datasets generated by the HCA and other groups could be leveraged in a wide variety of contexts, in both HD cytometry and single-cell SEQ
- --> by creating a platform and data-agnostic platform, we can efficiently incorporate various analysis tools, and particular integration tools, developed for cytometry (e.g. FlowSOM) and transcriptomics (e.g. Harmony, Liger, Seurat)
- By doing this, multiple approaches can be efficiently run and compared

--> in our case, focusing on methods that can scale to large datasets consisting of millions of cells (ref https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1850-9#Abs1)

Development
- Spectre for computational analysis
- Batch alignment/integration
- Paretobench and DR dev
- ICB paper example

Application
— examples in WNV/ZIKV/COVID etc

<br />

## Time-series clustering analysis for mapping dynamic immune responses

---

- Another major focus is the incorporation of time-series analysis approaches, as this is an under-developed field, but critical in many areas on biology

Development
- ChronoClust
- TrackSOM

Application
— examples in COVID etc

<br />

## Spatial analysis of high-dimensional imaging data

---

- Spatial analysis platform --> integrated with Spectre
- Improved cell segmentation workflows
- Spatial statistics and neighbourhood analysis
— examples in TBC

<br />
