---
layout: post
title: ASI Systems Immunology Hackathon (13-14 September 2022)
---

![](https://user-images.githubusercontent.com/11766139/173987029-1b2ba536-22a9-413b-82d2-0342a7afd3d9.jpeg](https://www.immunology.org.au/files/branch_meetings/Systems_Immunology_SIG_Hackathon_Flyer_8th.jpg)

<p>
<b>
<span style="font-size: 17px">
ASI Systems Immunology Hackathon (13-14 September 2022)
</span>
</b>
</p>

DATE: 13 to 14 September 2022
TIME: 11am on 13/9/22 to 2.45pm on 14/9/22
VENUE: Griffith University, Gold Coast Campus

**More info and rego: https://www.immunology.org.au/events/2022-ASI-Systems-Immunology-SIG-Hackathon/**

Recent advances in the simultaneous detection of the transcriptome and the proteome using single cell technologies have revolutionised our capacity to understand single cell types and states. This is particularly notable in the field of immunology where decades of flow cytometry technical and computational advances are merging with single cell RNA-seq data. However, many challenges in this new generation of multi-omics data exist in terms of pre-processing, QC, batch correction, cluster identification, feature selection, and suitable user interfaces for non-computational scientists to analyse and visualise their data.

This joint ASI - OzSingleCell hackathon will bring together leaders in cytometry, single cell genomics but also others from disparate fields such as UX/UI engineers, AI/ML, deep learning, graph neural networks, computer vision or image preprocessing, causal inference, etc in a single venue to first understand the challenges in analysis of proteo-genomics, and then work together in small groups to tackle some of these.
Open questions that need answering (or help with) in Immunology include; which cell types are likely to be interacting based on ligand-receptor pairs? how can I determine which are the minimal FACS markers to extract a population identified with scRNA+Protein-seq? Can I light up which cells have a particular pathway (e.g. active TCR signalling), machinery (e.g. antigen processing machinery) or other property? Can I merge FACS and single cell data together to better understand each? Can I more easily interrogate single cell data as a non-coder?

The hope is that by spending time in a room together, new collaborations, connections and tools will emerge that leverage Australia‚Äôs capacity in multi-omics analysis and build on our long legacy in immunology and genomics.

The intended outcomes for this workshop are:

- Seeding of new projects and collaborations
- Publication of a Grand Challenges summary
- New tools for visualisation and/or analysis of multi-omics data

**REGISTER BEFORE AUG 13TH 23:59 AEST**
üö© Please register now as there will be a cap of 50 participants to allow for interactions in small groups.

Registrations include:
‚úîÔ∏è Participation in the Hackathon
‚úîÔ∏è Lunches
‚úîÔ∏è Dinner
